server/src/main/java/org/opensearch/telemetry/tracing/listener/TraceableActionListener.java

```diff
@@ -0,0 +1,64 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * The OpenSearch Contributors require contributions made to
+ * this file be licensed under the Apache-2.0 license or a
+ * compatible open source license.
+ */
+
+package org.opensearch.telemetry.tracing.listener;
+
+import org.opensearch.common.util.FeatureFlags;
+import org.opensearch.core.action.ActionListener;
+import org.opensearch.telemetry.tracing.Span;
+
+import java.util.Objects;
+
+/**
+ * Tracer wrapped {@link ActionListener}
+ * @param <Response> response.
+ */
+public class TraceableActionListener<Response> implements ActionListener<Response> {
+
+    private final ActionListener<Response> delegate;
+    private final Span span;
+
+    /**
+     * Constructor.
+     * @param delegate delegate
+     * @param span span
+     */
+    private TraceableActionListener(ActionListener<Response> delegate, Span span) {
+        Objects.requireNonNull(delegate);
+        Objects.requireNonNull(span);
+        this.delegate = delegate;
+        this.span = span;
+    }
+
+    /**
+     * Factory method.
+     * @param delegate delegate
+     * @param span span
+     * @return action listener
+     */
+    public static ActionListener create(ActionListener delegate, Span span) {
+        if (FeatureFlags.isEnabled(FeatureFlags.TELEMETRY)) {
+            return new TraceableActionListener(delegate, span);
+        } else {
+            return delegate;
+        }
+    }
+
+    @Override
+    public void onResponse(Response response) {
+        span.endSpan();
```

Comment:
Same as other wrappers, we need to tract the time delegate spend on onResponse, right?

Reply:
one scenarion - 

method1() {
method2(new ActionListener(Span1))
}

method2(ActionListener) {
  method3(new ActionListener(span2))
}

method3(ActionListener listener){
listener.onResponse();
}

Spans will be closed in order Span1 and then Span2 but shouldn't ideally be Span1 took the maximum time?

Reply:
Correct, the problem is that right now, this piece of code

```
listener.onResponse();
```

is not included in any span time frame, and it probably has the most important computational part. One possible option could be instrumenting the onResponse():

```
ActionListener listener = new ActionListener() {
    void onResponse() {
      try(SpanScope span = tracer.spanInScope(...)) {
         ....
      }
    }
}

method3(ActionListener listener){
   listener.onResponse();
}
```

It also has trouble because  `onResponse` would not have the right parent. Which brings us to following implementation:

```
 @Override
    public void onResponse(Response response) {
        try(SpanScope scope= tracer.spanInScope(...)) {
             listener.onResponse();
        finally {
            span.endSpan();
        }
    }
```

Reply:
@reta, It still doesn't solve the issue of parent being closed before child (Span1::endTime > Span2::endTime). I think we need to have separate handling for Traceable delegates.

```
if(delegate instanceof TraceableActionListener){
    span.endSpan();
    delegate.onResponse(response);
}else{
   try {
        delegate.onResponse(response);
    }finally {
         span.endSpan();
    }
}
```

What do you think about this?

Reply:
@Gaganjuneja I think in this example the use of tracing wrappers is not correct: the `ActionListener` should be wrapped only once when created. 

> @reta, It still doesn't solve the issue of parent being closed before child (Span1::endTime > Span2::endTime).

This is not necessarily the  problem (very typical in async flows), but we should probably avoid layered `ActionListener` wrappers.

Reply:
@reta, Let's keep it simple for now and follow the approach you suggested initially.

```
try {
    delegate.handleResponse(response);
} finally {
        span.endSpan();
}
```

Reply:
> This is not necessarily the problem (very typical in async flows), but we should probably avoid layered ActionListener wrappers.

That is hard to restrict.

Reply:
> > This is not necessarily the problem (very typical in async flows), but we should probably avoid layered ActionListener wrappers.
> 
> That is hard to restrict.

We shouldn't (at this point), there only effect of end time, I will try to look more into it, but the missing instrumentation of the callback (like onResponse etc) is more serious - we loosing important execution information.